core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java

```diff
@@ -367,6 +368,50 @@ public void testPreferNodes() throws InterruptedException, IOException {
             terminate(threadPool);
         }
     }
+    
+    public void testFairSessionIdPreferences() throws InterruptedException, IOException {
+        // Ensure that a user session is re-routed back to same nodes for
+        // subsequent searches and that the nodes are selected fairly i.e.
+        // given identically sorted lists of nodes across all shard IDs
+        // each shard ID doesn't pick the same node.
+        TestThreadPool threadPool = null;
+        ClusterService clusterService = null;
+        try {
+            threadPool = new TestThreadPool("testFairSessionIdPreferences");
+            clusterService = ClusterServiceUtils.createClusterService(threadPool);
+            final String indexName = "test";
+            final int numShards = 4;
+            ClusterServiceUtils.setState(clusterService,
+                    ClusterStateCreationUtils.stateWithAssignedPrimariesAndOneReplica(indexName, numShards));
```

Comment:
So with random num shards and replicas I'm going to assume for this balancing test that shard assignments are optimal i.e. num nodes = shards x replicas (or only one Lucene index on each node). Otherwise I can't reason simply about how "bunched up" replicas are on each node and the expected number of nodes servicing a search